### DISCUSSION TITLE
Issues with crawlWrap function applied to temporally irregular GPS data

### DISCUSSION THREAD

**Discussion by NaimaDalpiaz** [bmcclintock/momentuHMM#152] [Q&A] [answered]
I am Naima Dalpiaz, a MSc student in Evolutionary Biology. I am using the momentuHMMpackage to work with Eurasian lynx telemetry data, obtained from GPS collars. I have encountered issues in the application of the crawlWrap function. 

This is an example of the dataset that I am using, with the same structure: 

[lynx_example.csv](https://github.com/bmcclintock/momentuHMM/files/9459695/lynx_example.csv)

The coordinate system is EPSG: 3035, and the dataset includes empty locations (as NAs). 
I used the crawlWrap function to regularize the data since the time steps in the original data are not regular. Since I only have 3 locations per day, I have chosen a time-step of 8 hours, leading to this code: 

crwOut_lynx<-crawlWrap(obsData=lynx_example,fixPar=c(1,1,NA,NA),Time.name="time",timeStep="8 hours", 
                  attempts=100, retryFits=20)

When trying to run this function I receive the following error message:

"Error in crawlWrap(obsData = lynx_example, fixPar = c(1, 1, NA, NA), Time.name = "time",  : 
  crawl::crwMLE failed for all individuals.  Check crawl::crwMLE arguments and/or consult crawl documentation."

I have tried running this with different theta values and putting the err.model argument as it is in the example shown in the help window. I left all other arguments as default. I have tried applying the possible solutions described [here](https://github.com/bmcclintock/momentuHMM/discussions/121), but it still gives me the same message. 

Do you know what might be causing this error?

I have also tried to run the Grey Seal script provided in the vignette of the package as an example of how to deal with temporally irregular data and how to implement biased movements relative to activity centers, which is what seemed closest to my data. By running the script I was able to regularize the data at 2 hr time-steps through the crawlWrap function, but then I wasn’t able to find cluster centroids corresponding to activity centers. When running the tsclust function, I received the following error message: 
" Error in summary.connection(connection) : invalid connection"

This error comes out even though I have been running the exact same script and data provided in the vignette. I tried some potential solutions as mentioned [here](https://stackoverflow.com/questions/64519640/error-in-summary-connectionconnection-invalid-connection) and [here](https://www.anycodings.com/questions/error-in-summaryconnectionconnection-invalid-connection-when-running-geolift-example),  but it still didn't work.

Do you know what could be causing this?

I am using the most recent R version (R version 4.2.1 (2022-06-23 ucrt), in a platform x86 64-w64-mingw32).

EDIT: I realized that the dataset I shared here doesn't have any missing values in the coordinates, which is the case for the data that I am dealing with. I thus created a different file that also includes 2 observations with missing values in the coordinates:
[lynx_ex.csv](https://github.com/bmcclintock/momentuHMM/files/9468399/lynx_ex.csv)

Thank you very much for your time. I look forward to receiving your feedback.

Naima Dalpiaz

**Comment by Zehidul-EcoMove** [comment]
Hi,
I have done a similar analysis for tigers using the crawl function for irregular data. Can you send another example file of your data with a column for date and time.  The "time" column in your above dataset seems to have no time.

**Reply by NaimaDalpiaz** [reply]
Hi,
Thank you for pointing that out, I had converted to POSIXct format incorrectly. 
Here is the file where there should be both date and time in the "time" column :
[lynx_example.csv](https://github.com/bmcclintock/momentuHMM/files/9459671/lynx_example.csv)
(I have also changed the file in the original question)
The structure is the following:
> str(lynx_example)
'data.frame':	45 obs. of  5 variables:
 $ ID                 : int  745 745 745 745 745 745 745 745 745 745 ...
 $ time               : POSIXct, format: "2022-04-01 20:00:00" "2022-04-02 04:00:00" "2022-04-03 00:00:00" "2022-04-03 20:00:00" ...
 $ gps_data_animals_id: int  585705 585707 585709 585711 585713 585715 585717 585719 585721 585723 ...
 $ x                  : num  4692894 4692892 4692787 4693285 4692930 ...
 $ y                  : num  2515993 2515986 2513092 2514720 2513106 ...

The error with the crawlWrap function is still the same.

**Comment by Zehidul-EcoMove** [comment]
Hi,
The first thing I did is made a new column for each date and time in the CSV file (attached below) and then run the codes provided.  
[lynx_example.csv](https://github.com/bmcclintock/momentuHMM/files/9460878/lynx_example.csv)

All the best!!


library(ggplot2)
library(momentuHMM)
library(sf)
library(sp)
library(mitools)
library(dplyr)
library(conicfit)
library(ggmap)


setwd("C:/Users/gis/Desktop/lynx")
HMM <- read.csv("lynx_example.csv", header = T)
head(HMM)


# add DateTime
HMM$DateTime <- as.POSIXct(lubridate::dmy(HMM$date) + lubridate::hms(HMM$time))
HMM$DateTime<-as.POSIXct(HMM$DateTime, format="%Y-%m-%d %H:%M:%0S")


# Keep relevant columns: ID, datetime, longitude, latitude and covariates (if you have any) in this order
names(HMM)
HMM <- HMM[, c(1,7, 5, 6)]
names(HMM)
colnames(HMM) <- c("ID", "time", "x", "y")
str(HMM)

#if you have covarites
#HMM <- HMM[, c(3, 18, 5, 6, 13, 14, 15, 17)]

#with covariates
#colnames(HMM) <- c("ID", "time", "lon", "lat", "dist.road", "elev", "temp","habitat")

#HMM$ID <- as.factor(HMM$ID)
#HMM$habitat <- as.factor(HMM$habitat)




# Project to UTM (if your data in latitude and longitude) or skip this
#llcoord <- st_as_sf(HMM[, c("lon", "lat")], coords = c("lon", "lat"), 
#                   crs = CRS("+proj=longlat +datum=WGS84"))

#utmcoord <- st_transform(llcoord, crs = CRS("+proj=utm +zone=44 +datum=WGS84"))

# Add Easting-Northing to data (in km)
#HMM[, c("x", "y")] <- st_coordinates(utmcoord)/1000



####------------------------ regularize track----------------------####
####-----------------------multiple imputation---------------------####

##fitting the crawl model
crw_out <- crawlWrap(obsData = HMM, timeStep = "8 hours", 
                     Time.name = "time", coord = c("x", "y"))



## plot the 'crwData'
plot(crw_out,
     animals = NULL,
     compact = FALSE,
     ask = FALSE,
     crawlPlot = FALSE)



### create momentuHMMData object from crwData object to calculate SL and TA without any covariates
mHMM <- prepData(data = crw_out)

##prep data for hmm with covariates (example)
#mHMM <- prepData(data = crw_out, covNames = c("dist.road", "elev", "temp", "habitat"))

###add column time of day (hour)
mHMM$hour <- as.integer(strftime(mHMM$time, format = "%H", tz="GMT"))

# Plot histogram of step lengths
hist(mHMM$step)


# Plot histogram of turning angles
hist(mHMM$angle, breaks = seq(-pi, pi, length = 15))


#Indices of steps of length zero
whichzero <- which(mHMM$step == 0)

# Proportion of steps of length zero in the data set, to include zero mass if there are zeros in sl
length(whichzero)/nrow(mHMM)

#acf 
acf(mHMM$step[!is.na(mHMM$step)],lag.max=720)

# plot prepped dataframe for each individual seperately
plot(mHMM) 

#save data in csv from later use
write.csv(mHMM,"regulartrack_8h.csv", row.names = FALSE)

**Reply by NaimaDalpiaz** [reply]
Thank you very much for all your help, although I am still having trouble...
The file I sent you before should already have the datetime column, named "time". 
While using the file you sent I was able to fit the crawlWrap() function but then, when applying the plot() function I received the following error message:
"Error in plot.window(...) : values 'xlim' must be finite"
I still wasn't able to apply the crawlWrap function when using the data that I shared as minimal reproducible example. 
Also I realized that the dataset I shared doesn't have any missing values in the coordinates, which is the case for the data that I am dealing with. I thus created a different file that also includes 2 observations with missing values in the coordinates:
[lynx_ex.csv](https://github.com/bmcclintock/momentuHMM/files/9468399/lynx_ex.csv)
This file should already have the correct projection. Since I wasn't able to do direcly the projection on a dataset with missing values in the coordinates I first removed the missing values, did the projection to epsg:3035, and then added again the observations with the missing coordinate values to the resulting dataframe. 
The result is the file attached to this message. 
I still get the same error message when trying to run the crawlWrap:
"Error in crawlWrap(obsData = lynx_ex, timeStep = "8 hours", Time.name = "time",  : 
  crawl::crwMLE failed for all individuals.  Check crawl::crwMLE arguments and/or consult crawl documentation"

Do you know what might be causing this?
Thank you again!

**Reply by Zehidul-EcoMove** [reply]
What I will suggest is to make two different columns: one for the date and another for the time in the CSV file. This code joins both date and time in R

HMM$DateTime <- as.POSIXct(lubridate::dmy(HMM$date) + lubridate::hms(HMM$time))
HMM$DateTime<-as.POSIXct(HMM$DateTime, format="%Y-%m-%d %H:%M:%0S")

And for interpolation of missing data through the crawl package as far as I know you don't have to have the row of missing locations. Your final CSV file shouldn't have the missing location. However, the code below will interpolate the locations.
Also, the sequence of columns should be the same as given in the above code.

crw_out <- crawlWrap(obsData = HMM, timeStep = "8 hours", Time.name = "time", coord = c("x", "y"))


Please try this way
Thanks

**Reply by NaimaDalpiaz** [reply]
From what I understand the momentuHMM should be able to deal with observations missing at random, but I still get the same error even when eliminating from the dataset all rows that thave NAs. I also have tried separating date and time and joining them after, as you suggested.

**Comment by bmcclintock** [comment, marked as answer]
> I am Naima Dalpiaz, a MSc student in Evolutionary Biology. I am using the momentuHMMpackage to work with Eurasian lynx telemetry data, obtained from GPS collars. I have encountered issues in the application of the crawlWrap function.
> 
> This is an example of the dataset that I am using, with the same structure:
> 
> [lynx_example.csv](https://github.com/bmcclintock/momentuHMM/files/9459695/lynx_example.csv)
> 
> The coordinate system is EPSG: 3035, and the dataset includes empty locations (as NAs). I used the crawlWrap function to regularize the data since the time steps in the original data are not regular. Since I only have 3 locations per day, I have chosen a time-step of 8 hours, leading to this code:
> 
> crwOut_lynx<-crawlWrap(obsData=lynx_example,fixPar=c(1,1,NA,NA),Time.name="time",timeStep="8 hours", attempts=100, retryFits=20)
> 
> When trying to run this function I receive the following error message:
> 
> "Error in crawlWrap(obsData = lynx_example, fixPar = c(1, 1, NA, NA), Time.name = "time", : crawl::crwMLE failed for all individuals. Check crawl::crwMLE arguments and/or consult crawl documentation."
> 
> I have tried running this with different theta values and putting the err.model argument as it is in the example shown in the help window. I left all other arguments as default. I have tried applying the possible solutions described [here](https://github.com/bmcclintock/momentuHMM/discussions/121), but it still gives me the same message.
> 
> Do you know what might be causing this error?
> 
> I am using the most recent R version (R version 4.2.1 (2022-06-23 ucrt), in a platform x86 64-w64-mingw32).
> 
> EDIT: I realized that the dataset I shared here doesn't have any missing values in the coordinates, which is the case for the data that I am dealing with. I thus created a different file that also includes 2 observations with missing values in the coordinates: [lynx_ex.csv](https://github.com/bmcclintock/momentuHMM/files/9468399/lynx_ex.csv)
> 
> Thank you very much for your time. I look forward to receiving your feedback.
> 
> Naima Dalpiaz

Please provide a complete minimal reproducible example. This should be a snippet of code that can be copied, pasted, and sourced by others from start to finish without any modification, including loading of the data (lynx_example.csv), any data preparation steps (e.g. adding projection or time stamp information), and anything else that you did up until the step that produces the error.

**Reply by NaimaDalpiaz** [reply]
Today I was eventually able to fit the crawlWrap function by starting again from the raw data and redoing the data preparation, including projection and time stamp, although this only works if I remove all NAs in the coordinates in the prepared data. Is it possible to use data with locations missing at random with this function ?
Thank you very much!

**Reply by bmcclintock** [reply]
Locations can most certainly be `NA`. `momentuHMM::crawlWrap` is just a wrapper function for `crawl::crwMLE` and `crawl::crwPredict` for fitting/predicting multiple individuals in parallel, so I suggest you consult the {crawl} help files (e.g. `?crawl::crwMLE`) and this [{crawl} workshop material](https://jmlondon.github.io/crawl-workshop/index.html) on data preparation, model specification, and prediction. Please note that I can't help you any further without a minimal reproducible example.

**Reply by NaimaDalpiaz** [reply]
I tried again also with missing locations included and I'm not sure what changed but now it works! 
Thank you very much for your patience and help, and for sending me the link to the material.

Here is a minimal reproducible example that I applied to this data [lynx_mre.csv](https://github.com/bmcclintock/momentuHMM/files/9526415/lynx_mre.csv) and that worked.


```
# LOAD PACKAGES
library(momentuHMM)
library(sf)

# set working directory
setwd("C:\\Users\\Naima\\Desktop\\lince\\Data_analisi\\New_files")
# load file
lynx <- read.csv("C:/Users/Naima/Desktop/lince/Data_analisi/New_files/lynx_mre.csv")

# set the acquisition_time column (which has datetime data) in format POSIXct
lynx$acquisition_time <- as.POSIXct(lynx$acquisition_time, format=c("%Y-%m-%d %H:%M:%OS"))

# Put columns in the correct order (ID, datetime, longitude, latitude) and change the name
names(lynx)
lynx <- lynx[, c(1,2, 4, 3)]
colnames(lynx) <- c("ID", "time", "x", "y")
str(lynx)

# Project to UTM (because my data is in long-lat)

# since missing values in the coordinates are not allowed to create a spatial object and do the projection I first remove The 
# rows with NAs in the coordinates and then add them back after having done the projection.

# Remove observations with NA in the coordinates
lynx_NONA <- lynx[complete.cases(lynx$x),]

# Do the projection
long_lat_coord <- st_as_sf(lynx_NONA[, c("x", "y")], coords = c("x", "y"),
                    crs = CRS("+proj=longlat +datum=WGS84"))
utm_coord <- st_transform(long_lat_coord, crs = CRS("+proj=utm +zone=34 +datum=WGS84"))
lynx_NONA[, c("x", "y")] <- st_coordinates(utm_coord)/1000

# Now I put back in the observations with NAs in the coordinates 
lynx_NA <- lynx[is.na(lynx$x),]
lynx_tot <- rbind(lynx_NONA, lynx_NA)
str(lynx_tot)

# Fit the crawlWrap function to regularize track
crw_lynx <- crawlWrap(obsData = lynx_tot, timeStep = "8 hours",
                        Time.name = "time", coord = c("x", "y"))

plot(crw_lynx,
     animals = NULL,
     compact = FALSE,
     ask = FALSE,
     crawlPlot = FALSE)
```

**Comment by bmcclintock** [comment]
> I have also tried to run the Grey Seal script provided in the vignette of the package as an example of how to deal with temporally irregular data and how to implement biased movements relative to activity centers, which is what seemed closest to my data. By running the script I was able to regularize the data at 2 hr time-steps through the crawlWrap function, but then I wasn’t able to find cluster centroids corresponding to activity centers. When running the tsclust function, I received the following error message: " Error in summary.connection(connection) : invalid connection"
> 
> This error comes out even though I have been running the exact same script and data provided in the vignette. I tried some potential solutions as mentioned [here](https://stackoverflow.com/questions/64519640/error-in-summary-connectionconnection-invalid-connection) and [here](https://www.anycodings.com/questions/error-in-summaryconnectionconnection-invalid-connection-when-running-geolift-example), but it still didn't work.
> 
> Do you know what could be causing this?

Please create a separate post for your question about the grey seal example, but, before you do, make sure you have tried running the grey seal example in a fresh R session after updating packages to their most recent versions, reinstalling {momentuHMM}, and restarting your machine.
